packages/eslint-plugin/tests/rules/sort-lifecycle-methods/cases.ts

```diff
@@ -0,0 +1,190 @@
+    messageId: lifecycleMethodsNotSorted,
+  }),
+  convertAnnotatedSourceToFailureCase({
+    description: 'ngAfterViewChecked() is declared after ngOnDestroy()',
+    annotatedSource: `
+      @Component()
+      class Test {
+        ngOnChanges(): void {}
+        ngOnInit(): void {}
+        ngAfterContentInit(): void {}
+        ngAfterContentChecked(): void {}
+        ngOnDestroy(): void {}
+        ngAfterViewChecked(): void {}
+        ~~~~~~~~~~~~~~~~~~
+        doSomething(): void {}
+      }
+      `,
+    messageId: lifecycleMethodsNotSorted,
+  }),
+  convertAnnotatedSourceToFailureCase({
+    description: 'ngOnDestroy() is declared after ngAfterContentInit()',
+    annotatedSource: `
+      @Component()
+      class Test {
+        ngDoCheck(): void {}
+        ngAfterContentInit(): void {}
+        ngOnDestroy(): void {}
+        ngAfterContentChecked(): void {}
+        ~~~~~~~~~~~~~~~~~~~~~
+        ngAfterViewChecked(): void {}
+        doSomething(): void {}
+        doSomethingElse(): void {}
+      }
+      `,
+    messageId: lifecycleMethodsNotSorted,
+  }),
+  convertAnnotatedSourceToFailureCase({
+    description: 'ngOnChanges() is declared after ngOnInit()',
+    annotatedSource: `
+      @Component()
+      class Test {
+        ngOnInit(): void {}
+        ngOnChanges(): void {}
+        ~~~~~~~~~~~
+       }
+      `,
+    messageId: lifecycleMethodsNotSorted,
+  }),
+  convertAnnotatedSourceToFailureCase({
+    description: 'doSomething() is declared before ngOnInit()',
+    annotatedSource: `
+      @Component()
+      class Test {
+        doSomething(): void {}
+        ~~~~~~~~~~~
+        ngOnInit(): void {}
+       }
+      `,
+    messageId: nonLifecycleMethodBeforeLifecycleMethod,
+  }),
+  convertAnnotatedSourceToFailureCase({
+    description: 'doSomething() is declared before ngDoCheck()',
+    annotatedSource: `
+      @Component()
+      class Test {
+        ngOnChanges(): void {}
+        ngOnInit(): void {}
+        doSomething(): void {}
+        ~~~~~~~~~~~
+        ngDoCheck(): void {}
+        ngAfterContentInit(): void {}
+        ngAfterContentChecked(): void {}
+        ngAfterViewInit(): void {}
+        ngAfterViewChecked(): void {}
+        ngOnDestroy(): void {}
+        doSomethingElse(): void {}
+      }
+      `,
+    messageId: nonLifecycleMethodBeforeLifecycleMethod,
+  }),
+  convertAnnotatedSourceToFailureCase({
+    description: 'lifecycle hook is declared before constructor()',
+    annotatedSource: `
+      @Component()
+      class Test {
+        ngOnChanges(): void {}
+        ngOnInit(): void {}
+        ngDoCheck(): void {}
+        ngAfterContentInit(): void {}
+        ngAfterContentChecked(): void {}
+        ngAfterViewInit(): void {}
+        ngAfterViewChecked(): void {}
+        ngOnDestroy(): void {}
+        ~~~~~~~~~~~
+        constructor() {}
+        doSomethingElse(): void {}
+      }
+      `,
+    messageId: lifecycleMethodBeforeConstructor,
+  }),
+];
```

Comment:
What's the expected behavior if more than one lifecycle hook is in the incorrect order?

Reply:
@sc0tt5 Should we flag all inversions? Or can we find a more optimal & deterministic way to flag methods?

Reply:
I'm assuming that with the following example...
```ts
        ngOnInit(): void {}
        ngOnChanges(): void {}
        ngAfterContentInit() {}
        ngDoCheck() {}
```
...this rule will report twice: first on `ngOnChanges` and then on `ngDoCheck`. If my assumption is correct, I think this approach is appropriate.

Reply:
Just want to clarify, in the example

```javascript
        ngOnInit() {}
        ngOnChanges() {}
        ngAfterContentInit() {}
        ngDoCheck() {}
```

There's two inversion pairs `(ngOnInit, ngOnChanges)` and `(ngAfterContentInit, ngDoCheck)`. The agreed rule is for each inversion pair, we flag the second element in the pair.

Reply:
Thank you for clarifying. Makes sense to me 🙂.

One thing to keep in mind with this approach and if there will be an auto-fixer for this rule in the future... The ESLint fix "process will repeat up to 10 times, or until no more fixable problems are found," as documented [here](https://eslint.org/docs/latest/extend/custom-rules#applying-fixes).

Should be okay for this rule because there are only 8 Angular lifecycle events, at least for now.

Reply:
If possible, maybe let's play around with test cases that check for multiple errors? See [here](https://github.com/angular-eslint/angular-eslint/blob/main/packages/utils/src/convert-annotated-source-to-failure-case.ts#L77)

Reply:
@sc0tt5 What is your opinion on the inversion based flagging?